### PR TITLE
Rework nth() on top of vctrs

### LINKED ISSUE #3789: Lead/Lag behavior with nested column
I'm not sure if this is an intentional design, but I believe that lead and lag do not behave appropriately with nested columns.  

``` r
library(tidyverse)
#> Warning: package 'tidyverse' was built under R version 3.4.4
#> -- Attaching packages -------------------------------- tidyverse 1.2.1 --
#> v ggplot2 3.0.0     v purrr   0.2.4
#> v tibble  1.4.2     v dplyr   0.7.6
#> v tidyr   0.8.1     v stringr 1.3.1
#> v readr   1.1.1     v forcats 0.2.0
#> Warning: package 'ggplot2' was built under R version 3.4.4
#> Warning: package 'tibble' was built under R version 3.4.4
#> Warning: package 'tidyr' was built under R version 3.4.4
#> Warning: package 'purrr' was built under R version 3.4.2
#> Warning: package 'dplyr' was built under R version 3.4.4
#> Warning: package 'stringr' was built under R version 3.4.4
#> -- Conflicts ----------------------------------- tidyverse_conflicts() --
#> x dplyr::filter() masks stats::filter()
#> x dplyr::lag()    masks stats::lag()

dt <- 
  tibble::tribble(
  ~id, ~a, ~b,
   1L, 0L, 0L,
   2L, 0L, 1L,
   3L, 1L, 0L,
   4L, 1L, 1L
  )

# doesn't work
dt %>% 
  nest(-id) %>% 
  mutate(prior = lag(data, default = first(data)))
#> Warning: package 'bindrcpp' was built under R version 3.4.4
#> Error in mutate_impl(.data, dots): Column `prior` must be length 4 (the number of rows) or one, not 5

# works
dt %>% 
  nest(-id) %>% 
  mutate(prior = lag(data, default = list(first(data))))
#> # A tibble: 4 x 3
#>      id data             prior           
#>   <int> <list>           <list>          
#> 1     1 <tibble [1 x 2]> <tibble [1 x 2]>
#> 2     2 <tibble [1 x 2]> <tibble [1 x 2]>
#> 3     3 <tibble [1 x 2]> <tibble [1 x 2]>
#> 4     4 <tibble [1 x 2]> <tibble [1 x 2]>
```
Why do I need to encapsulate with list?

Edit: On second thought, perhaps this has to do with how `first` works?

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #3789

``` r
library(tidyverse)

dt <- tribble(
    ~id, ~a, ~b,
    1L, 0L, 0L,
    2L, 0L, 1L,
    3L, 1L, 0L,
    4L, 1L, 1L
  )

dt %>% 
  group_nest(id) %>% 
  mutate(prior = lag(data, default = first(data)))
#> # A tibble: 4 x 3
#>      id           data          prior
#>   <int> <list<df[,2]>> <list<df[,2]>>
#> 1     1        [1 × 2]        [1 × 2]
#> 2     2        [1 × 2]        [1 × 2]
#> 3     3        [1 × 2]        [1 × 2]
#> 4     4        [1 × 2]        [1 × 2]
```

<sup>Created on 2019-11-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

Also: 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
first(iris)
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
first(list(1, 2, 3))
#> [[1]]
#> [1] 1

last(iris)
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1            5         3.6          1.4         0.2  setosa
last(list(1, 2, 3))
#> [[1]]
#> [1] 3

lag(head(iris, 7), 1)
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1           NA          NA           NA          NA    <NA>
#> 2          5.1         3.5          1.4         0.2  setosa
#> 3          4.9         3.0          1.4         0.2  setosa
#> 4          4.7         3.2          1.3         0.2  setosa
#> 5          4.6         3.1          1.5         0.2  setosa
#> 6          5.0         3.6          1.4         0.2  setosa
#> 7          5.4         3.9          1.7         0.4  setosa
lag(list(1, 2, 3), 1)
#> [[1]]
#> [1] NA
#> 
#> [[2]]
#> [1] 1
#> 
#> [[3]]
#> [1] 2
```

<sup>Created on 2019-11-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

`lag()` and `lead()` expose `NA` as the default value for `default=` but it then gets `vec_cast()`'ed to the appropriate `vec_ptype()`

Hence the `default_missing()` function, which was never exported, has been removed.

### CODE REVIEW

File: tests/testthat/test-nth-value.R [hunk 1]
```diff
@@ -3,9 +3,9 @@ context("Nth value")
 test_that("nth works with lists", {
   x <- list(1, 2, 3)
 
-  expect_equal(nth(x, 1), 1)
-  expect_equal(nth(x, 4), NULL)
-  expect_equal(nth(x, 4, default = 1), 1)
+  expect_equal(nth(x, 1), vec_slice(x, 1))
```
Comment (hadley) [inline tests/testthat/test-nth-value.R]: I think you're better leaving these as constants so it's easier to understand when reading the tests

File: R/nth-value.R [hunk 2]
```diff
@@ -53,38 +53,20 @@ nth <- function(x, n, order_by = NULL, default = default_missing(x)) {
   }
 
   if (is.null(order_by)) {
-    x[[n]]
+    vec_slice(x, n)
```
Comment (hadley) [inline R/nth-value.R]: I'm not sure about this — I think `nth()` is equivalent to `[[` not `[`.

File: tests/testthat/test-nth-value.R (hunk 1 above)
Comment (romainfrancois) [reply tests/testthat/test-nth-value.R]: I don't know, `nth()` for a matrix would be natural as a one row matrix, same for data frame ...

Comment (hadley) [reply tests/testthat/test-nth-value.R]: That's good motivation, and I have read through related issue and I think I'm mostly convinced, but I'd like to think about it a bit more before we merge this.

Comment (hadley) [reply tests/testthat/test-nth-value.R]: Also maybe we should just put this in funs?

Comment (DavisVaughan) [reply tests/testthat/test-nth-value.R]: Is this where a `vec_get()` would be used then? Feels correct

**Comment by hadley** [comment]
I'm now convinced that this is the write approach, but would you mind contributing it to funs instead of dplyr?

**Comment by lionel-** [comment]
Why are we changing `first()` to be a wrapper around `[` rather than `[[`? I find this surprising.

I agree with Davis this is where `vec_get()` would come into play, though I'm not sure we should treat data frames rowwise in user-facing APIs. Should we discuss this at a high-level before changing semantics piecemeals in funs / dplyr?

Regarding backward compatibility, both these changes (non-extracting slice for lists, or rowwise slicing/extraction for data frames) would break many packages. Is the plan to move `first()` to funs and deprecate the dplyr version?

**Comment by hadley** [comment]
I don't think it will break many packages because I don't think many people use `first()` with list-columns. And confusingly, it already seems to behave like `[`?

``` r
library(dplyr, warn.conflicts = FALSE)
packageVersion("dplyr")
#> [1] '0.8.3'

df <- tibble(g = 1:3, y = list(1, 1:2, 1:3))
df %>% group_by(g) %>% summarise(y = first(y))
#> # A tibble: 3 x 2
#>       g y        
#>   <int> <list>   
#> 1     1 <dbl [1]>
#> 2     2 <int [2]>
#> 3     3 <int [3]>

df %>% group_by(g) %>% summarise(y = y[[1]])
#> Error: Column `y` must be length 1 (a summary value), not 2
df %>% group_by(g) %>% summarise(y = y[1])
#> # A tibble: 3 x 2
#>       g y        
#>   <int> <list>   
#> 1     1 <dbl [1]>
#> 2     2 <int [2]>
#> 3     3 <int [3]>
```

<sup>Created on 2019-12-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by lionel-** [comment]
My sense is that `first()` is also imported and used as a general vector accessor. In that case it's not just used inside dplyr verbs but also outside.

**Comment by lionel-** [comment]
`last()` is probably more problematic than `first()`, since `long_name[[length(long_name)]]` is more likely to make users reach for a wrapper than `long_name[[1]]`.

**Comment by DavisVaughan** [comment]
> And confusingly, it already seems to behave like [?

I think this has to do with hybrid eval. Replacing with the equivalent that uses standard eval works like `[[`

``` r
library(dplyr, warn.conflicts = FALSE)
packageVersion("dplyr")
#> [1] '0.8.3'

df <- tibble(g = 1:3, y = list(1, 1:2, 1:3))

df %>% group_by(g) %>% summarise(y = first(y))
#> # A tibble: 3 x 2
#>       g y        
#>   <int> <list>   
#> 1     1 <dbl [1]>
#> 2     2 <int [2]>
#> 3     3 <int [3]>

df %>% group_by(g) %>% summarise(y = first(y, default = dplyr:::default_missing(y)))
#> Error: Column `y` must be length 1 (a summary value), not 2

first(df$y)
#> [1] 1
```

**Comment by hadley** [comment]
Closing this PR; we'll implement in funs so we can preserve existing behaviour in dplyr, and figure out a long term deprecation strategy.

**Pull request closed without merging**